em-app/src/utils.rs

```diff
@@ -0,0 +1,208 @@
+                 ca_crl: Option<Arc<Crl>>
+) -> Result<Vec<u8>, String> {
+    let mut config = Config::new(Endpoint::Client, Transport::Stream, Preset::Default);
+    
+    config.set_rng(Arc::new(mbedtls::rng::Rdrand));
+    config.set_min_version(Version::Tls1_2).map_err(|e| format!("TLS configuration failed: {:?}", e))?;
+
+    if let Some(ca_cert_list) = ca_cert_list {
+        config.set_ca_list(ca_cert_list, ca_crl);
+        config.set_authmode(AuthMode::Required);
+    } else {
+        config.set_authmode(AuthMode::Optional);
+    }
+
+    let ssl = MbedSSLClient::new(Arc::new(config), true);
+    let connector = HttpsConnector::new(ssl);
+    let client = hyper::Client::with_connector(Pool::with_connector(Default::default(), connector));
+    let mut response = client.get(url).send().map_err(|e| format!("Failed downloading, error: {:?}, url: {}", e, url))?;
+
+    if response.status != hyper::status::StatusCode::Ok {
+        return Err(format!("Request failed, result: {:?}", response));
+    }
+    
+    let mut body = vec![];
+    response.read_to_end(&mut body).map_err(|e| format!("Failed reading body, error: {:?}", e))?;
+
+    Ok(body)
+}
+
+pub fn https_put(url: &str,
+                 body: Vec<u8>,
+                 ca_cert_list: Option<Arc<MbedtlsList<Certificate>>>,
+                 ca_crl: Option<Arc<Crl>>
+) -> Result<(), String> {
+    let mut config = Config::new(Endpoint::Client, Transport::Stream, Preset::Default);
+    
+    config.set_rng(Arc::new(mbedtls::rng::Rdrand));
+    config.set_min_version(Version::Tls1_2).map_err(|e| format!("TLS configuration failed: {:?}", e))?;
+
+    if let Some(ca_cert_list) = ca_cert_list {
+        config.set_ca_list(ca_cert_list, ca_crl);
+        config.set_authmode(AuthMode::Required);
+    } else {
+        config.set_authmode(AuthMode::Optional);
+    }
+
+    let ssl = MbedSSLClient::new(Arc::new(config), true);
+    let connector = HttpsConnector::new(ssl);
+    let client = hyper::Client::with_connector(Pool::with_connector(Default::default(), connector));
+    let result = client.put(url).body(body.as_slice()).send().map_err(|e| format!("Failed upload, error: {:?}, url: {}", e, url))?;
+
+    if result.status != hyper::status::StatusCode::Ok {
+        return Err(format!("Request failed, result: {:?}", result));
+    }
+    
+    Ok(())
+}
+
+const NONCE_SIZE : usize = 12;
+const TAG_SIZE : usize = 16;
+
+// Basic AES-256-GCM encrypt/decrypt utility functions.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct CredentialsEncryption {
+    pub key: String,
+}
+
+pub fn encrypt_buffer(body: &[u8], encryption: &CredentialsEncryption) -> Result<Vec<u8>, String>{
+    let key = encryption.key.from_hex().map_err(|e| format!("Failed deconding key as a hex string: {:?}", e))?;
+
+    let mut nonce = [0; NONCE_SIZE];
+    Rdrand.random(&mut nonce[..]).map_err(|e| format!("Could not generate andom nonce {}", e))?;
```

Comment:
There is a typo "deconding" on line 171 (also below in decrypt), and a typo "andom" here.

Reply:
thanks, fixed